### PR TITLE
Implementation of GatewayConnection

### PR DESCRIPTION
**Pull request by Lys4nder**
To Do:
- [x] Write Gmock Tests
- [x] Reimplement receive logic

### CODE REVIEW

File: tests/GatewayConnectionTests.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+public:
+    MockGatewayConnection(std::string ip, int32_t port) : GatewayConnection(ip, port) {}
+
+    MOCK_METHOD(void, SendMessage, (std::string message));
+    MOCK_METHOD(std::string, Receive, ());
+};
+
+using ::testing::_;
+using ::testing::Invoke;
+using ::testing::Return;
```
Comment (CristianIonutIonescu) [inline tests/GatewayConnectionTests.h:14]: why do we need these here?

File: src/GatewayConnection.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+void GatewayConnection::SendMessage(std::string message)
+{
+    // send message with the request, can be anything but an empty string
+    if (message.empty()) {
+        perror("Error: Can't send an empty string!");
+        exit(EXIT_FAILURE);
+    }
+    // convert the message to a const char * as to have the same number of bytes when sending the message
+    const char *msg = message.c_str();
+    int bytesSent = send(socket_, msg, strlen(msg), 0);
```
Comment (CristianIonutIonescu) [inline src/GatewayConnection.cpp:40]: why do you need an extra variable? can't message.c_str() be used in the send method?

File: include/GatewayConnection.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+#define BACKEND_SERVICE_INCLUDE_GATEWAYCONNECTION_H_
+
+#include <string>
+#include <netinet/in.h>
+
+class GatewayConnection {
+    public:
+        GatewayConnection(std::string ip, int32_t port);
+        std::string GetData();
+        ~GatewayConnection();
```
Comment (CristianIonutIonescu) [inline include/GatewayConnection.h:11]: destructor before GetData()

File: include/GatewayConnection.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#ifndef BACKEND_SERVICE_INCLUDE_GATEWAYCONNECTION_H_
+#define BACKEND_SERVICE_INCLUDE_GATEWAYCONNECTION_H_
+
+#include <string>
+#include <netinet/in.h>
+
+class GatewayConnection {
```
Comment (CristianIonutIonescu) [inline include/GatewayConnection.h:7]: what about rule of five/zero? Is this class copiable?

File: include/GatewayConnection.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    public:
+        GatewayConnection(std::string ip, int32_t port);
+        std::string GetData();
+        ~GatewayConnection();
+    private:
+        void SendMessage(std::string message);
+        std::string Receive();
+        std::string ip_;
+        int32_t port_;
```
Comment (CristianIonutIonescu) [inline include/GatewayConnection.h:16]: are these used outside of constructor?

File: include/GatewayConnection.h [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        int32_t socket_;
+        struct sockaddr_in server_address_;
```
Comment (CristianIonutIonescu) [inline include/GatewayConnection.h:18]: do you use this outside of constructor?

File: src/main.cc [hunk 7]
```diff
@@ -8,5 +9,9 @@ int main()
 
     std::cout << p.name() << std::endl;
```
Comment (CristianIonutIonescu) [inline src/main.cc:10]: do you still need this?

**Review by CristianIonutIonescu — CHANGES_REQUESTED** [review]
there's still work to do

File: src/GatewayConnection.cpp [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+void RTopServer::GatewayConnection::SendMessage(std::string message)
+{
+    // send message with the request, can be anything but an empty string
+    if (message.empty()) {
+        perror("Error: Can't send an empty string!");
+        exit(EXIT_FAILURE);
+    }
+    
+    int bytesSent = send(socket_, message.c_str(), strlen(message.c_str()), 0);
```
Comment (marisvali) [inline src/GatewayConnection.cpp:42]: Use message.length() or message.size(), they are O(1) operations. strlen(message.c_str()) is O(n).

File: src/GatewayConnection.cpp [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+{
+    //Combine SendMessage and Receive methods for SOLID and RAII
+    SendMessage("data");
+    return Receive();
+}
+
+
+RTopServer::GatewayConnection::~GatewayConnection()
+{
+    if (close(socket_) == -1) {
```
Comment (marisvali) [inline src/GatewayConnection.cpp:107]: You should also do shutdown() before close().

File: src/GatewayConnection.cpp [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+RTopServer::GatewayConnection::GatewayConnection(GatewayConnection&& other)
+{
+    socket_ = std::move(other.socket_);
+}
+
+RTopServer::GatewayConnection& RTopServer::GatewayConnection::operator=(GatewayConnection&& other)
+{
+    socket_ = std::move(other.socket_);
```
Comment (marisvali) [inline src/GatewayConnection.cpp:93]: std::move on an integer has no effect, you're not moving anything, just making a copy. std::move should be removed. But now the question is, is it ok to make a copy? Why yes/no?

File: src/GatewayConnection.cpp [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        return *this;
+    }
+
+    socket_ = other.socket_;
+    return *this;
+}
+
+RTopServer::GatewayConnection::GatewayConnection(GatewayConnection&& other)
+{
+    socket_ = std::move(other.socket_);
```
Comment (marisvali) [inline src/GatewayConnection.cpp:88]: std::move on an integer has no effect, you're not moving anything, just making a copy. std::move should be removed. But now the question is, is it ok to make a copy? Why yes/no?

**Review by marisvali — CHANGES_REQUESTED** [review]

File: tests/GatewayConnectionTests.h
Comment (enachebogdan) [inline tests/GatewayConnectionTests.h:1]: I've noticed you only test copy/move semantics, nothing relevant to the GatewayConnection functionality. You should find a way to test all of your functions.

File: tests/GatewayConnectionTests.h [hunk 12]
```diff
@@ -0,0 +1,97 @@
+#include <gmock/gmock.h>
+#include "../include/GatewayConnection.h"
+
+
+// Before running tests, make sure to start server: nc -l 127.0.0.1 8080
```
Comment (enachebogdan) [inline tests/GatewayConnectionTests.h:5]: Usually with unit tests we aim for an automatic execution, with no manual steps (starting this server). You should just use mock data to test the functionality.

If you need to use this server for your manual tests that's perfectly fine, it just shouldn't be in the unit test process.

File: include/gateway_connection.h [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+#include <netinet/in.h>
+
+namespace RTopServer {
+    class GatewayConnection {
+    public:
+        GatewayConnection(std::string ip, int32_t port);
+        GatewayConnection(const GatewayConnection& other) = delete;             // copy constructor
+        GatewayConnection& operator=(const GatewayConnection& other) = delete;  // copy assignment operator
+        GatewayConnection(GatewayConnection&& other);                           // move constructor
+        GatewayConnection& operator=(GatewayConnection&& other);                // move assignment operator
```
Comment (CristianIonutIonescu) [inline include/gateway_connection.h:14]: doesn't respect google style guide regarding comments

File: tests/proto_tests.cpp [hunk 14]
```diff
@@ -0,0 +1,50 @@
+#include <iostream>
+#include <gtest/gtest.h>
+#include "../build/proto_generated/data.pb.h"
+
+TEST(MemoryUsageTests, Test)
```
Comment (CristianIonutIonescu) [inline tests/proto_tests.cpp:5]: why do you name all your tests here? read what the parameters of TEST mean

Comment (Lys4nder) [reply tests/proto_tests.cpp:5]: this is not my code, this is from pull origin master

**Review by CristianIonutIonescu — CHANGES_REQUESTED** [review]
still missing google coding guidelines regarding class and method comments

File: include/gateway_connection.h [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+namespace RTopServer {
+    // used to connect via TCP to Gateway Server
+    // this uses client side implementation that needs to connect to an existing socket
+    // it needs to know the ip and port of the server in order to connect to it
+    class GatewayConnection {
+    public:
+        // this constructor handles connection to Gateway Server
+        // when creating the object, it needs to know the ip and port
+        // of the server
+        GatewayConnection(std::string ip, int32_t port);
```
Comment (CristianIonutIonescu) [inline include/gateway_connection.h:16]: I thought we agreed that the port should be an uint16_t, or I discussed in another PR?

**Review by CristianIonutIonescu — COMMENTED** [review]
last issue I saw

**Review by CristianIonutIonescu — APPROVED** [review]

**Pull request merged**